TLS creation/ renewal instructions (#172)
astronomer/docs: enterprise/next/06_manage-astronomer/09_renew-tls-cert.md

---
title: "Renew TLS Certificates on Astronomer Enterprise"
navTitle: "Renew a TLS Certificate"
description: "How to update and auto-renew your organization's TLS Certificate for Astronomer"
---

## Overview

Once you set up an TLS certificate for your Astronomer platform, you'll need to establish a process for periodically renewing the certificate. This can be done in one of two ways:

* **Automatic renewal**: Let's Encrypt provides a service which automatically renews your TLS certificate every 90 days. We recommend this option for smaller organizations where your DNS administrator and cluster administrator are either the same person or on the same team.

> * **Manual renewal**: Manual renewal looks very similar to the initial certificate creation process, except you additionally need to delete your current certificate and update your systems to begin using the new certificate. We recommend this method for large organizations that have their own protocols for issuing certificates.

Reviewer: ```suggestion
* **Manual renewal**: Manual renewal looks very similar to the initial certificate creation process, except you are replacing your existing certificate with a new certificate. We recommend this method for large organizations that have their own protocols for issuing certificates, and certificates that last long enough to justify manual update.
```

Author: I think the last clause is redundant (or implied if an org has their own practices for this), but I like your wording on the first part!

## Automatically Renew TLS Certificates Using Let's Encrypt

Let's Encrypt is a Certificate Authority that provides free, 90-day certificates using the ACME protocol. You can use the Cert Manager project for Kubernetes to automatically renew certificates.

1. Install the Kubernetes Cert Manager by following [the official installation guide](https://cert-manager.io/docs/installation/kubernetes/).

> 2. If you use AWS, grant Cert Manager access to your nodes by updating your `NodeInstanceRole` to include the following policies (if you don't use AWS, skip this step):

Reviewer: ```suggestion
2. If you use AWS, grant your nodes access to Route 53 by adding a role similar to this CloudFormation snippet to your nodes' Instance Profile (if you don't use AWS, skip this step):
```

Reviewer: "node", "Route 53", and "Instance Profile" are key words. This section below is technically a cloudformation snippet, which is the most likely way this will be consumed. If creating the role point-and-click in the AWS UI then they will need to perform the substitution of the hosted zone id (line 37).

> 2. If you use AWS, grant Cert Manager access to your nodes by updating your `NodeInstanceRole` to include the following policies (if you don't use AWS, skip this step):

Reviewer: regarding "skip this step" - they need to find some way to authenticate cert manager against their DNS provider.

```yaml
Type: "AWS::IAM::Role"
Properties:
  RoleName: instance-profile-role
  Policies:
    - PolicyName: instance-profile-policy
      PolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Action: route53:GetChange
            Resource: arn:aws:route53:::change/*
          - Effect: Allow
            Action:
              - route53:ChangeResourceRecordSets
              - route53:ListResourceRecordSets
            Resource: !Sub arn:aws:route53:::hostedzone/${HostedZoneIdLookup.HostedZoneId}
          - Effect: Allow
            Action: route53:ListHostedZonesByName
            Resource: '*'
  AssumeRolePolicyDocument:
    Version: "2012-10-17"
    Statement:
      - Effect: "Allow"
        Principal:
          Service:
            - "ec2.amazonaws.com"
        Action:
          - "sts:AssumeRole"
```
For more information on how to complete this setup, refer to the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/create-node-role.html).

> 3. Create a "ClusterIssuer" resource that declares how requests for certificates will be fulfilled. To do so, first create the a `clusterissuer.yaml` file with the following values:

Author: In @shmanu017's PR for a similar document, he had a single astronomer-cert.yaml file with the configuration from both ClusterIssuer.yaml certificate.yaml below. This differs from the setup Steven described in his doc (https://github.com/astronomer/labs/tree/master/tls-certificate) What's the best/right way to configure this?

Reviewer: the indentation below looks incorrect to me.

Author: @shmanu017 Can you confirm what this .yaml file should look like?

Reviewer 2: ```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
    name: letsencrypt-prod
spec:
    acme:
        email: <your-email>
        server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
        name: cert-manager-issuer-secret-key
    solvers:
        - selector: {}
    dns01:
        route53:
            region: <your-server-region>
          
```

```yaml

> apiVersion: cert-manager.io/v1
> kind: ClusterIssuer
> metadata:
>     name: letsencrypt-prod
> spec:
>     acme:
>         email: <your-email>
>         server: https://acme-v02.api.letsencrypt.org/directory
>     privateKeySecretRef:
>         name: cert-manager-issuer-secret-key
>     solvers:
>         - selector: {}
>         dns01:
>             route53:
>                 region: <your-server-region>

Reviewer: ```suggestion
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-prod
spec:
  acme:
    email: <your-email>
    server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: cert-manager-issuer-secret-key
    solvers:
    - selector: {}
      dns01:
        route53:
          region: <your-server-region>
```

```

> Then, create the cluster by running the following command:

Reviewer: ```suggestion
Then, create the ClusterIssuer by running the following command:
```

```sh
kubectl apply -f clusterissuer.yaml
```

> 4. Create a "Certificate" resource that declares the type of certificate you'll request from Let's Encrypt. To do so, first create the a `certificate.yaml` file, replacing `BASE_DOMAIN` with your organization's own base domain:

Reviewer: ```suggestion
4. Create a "Certificate" resource that declares the type of certificate you'll request from Let's Encrypt. To do so, first create the a `certificate.yaml` file, replacing `BASE_DOMAIN`:
```

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
    name: acme-crt
spec:
    secretName: astronomer-tls
    dnsNames:
        - BASE_DOMAIN
        - app.BASE_DOMAIN
        - deployments.BASE_DOMAIN
        - registry.BASE_DOMAIN
        - houston.BASE_DOMAIN
        - grafana.BASE_DOMAIN
        - kibana.BASE_DOMAIN
        - install.BASE_DOMAIN
        - prometheus.BASE_DOMAIN
        - alertmanager.BASE_DOMAIN
    issuerRef:
        name: letsencrypt-prod
        kind: ClusterIssuer
        group: cert-manager.io
```
Then, create the cluster by running the following command and waiting a few minutes:
```sh
kubectl apply -f certificate.yaml
```
5. Ensure that the certificate was created by running:
```sh
kubectl get certificates
```

> ## Manually Renew TLS certificates

Author: Inferred setup based on my general knowledge.

> Larger organizations with dedicated security teams will likely have their own processes for requesting and renewing TLS certificates. Regardless, there are specific steps you have to complete for Astronomer when renewing TLS certificates:

Reviewer: here you say "process" which I prefer to saying "protocol", which is said elsewhere. I think "protocol" is a bit strange sounding if referring to a business process.

1. Delete your current TLS certificate by running the following command:
```sh
kubectl delete secret astronomer-tls
```
2. Follow the instructions for requesting an TLS certificate from your organization's security team as described in [Step 4: Configure TLS](https://www.astronomer.io/docs/enterprise/stable/install/aws/install-aws-standard#step-4-configure-tls). The linked guide is for setting up with AWS, but this step is the same regardless of which service you use.

> 3. Restart your Houston, nginx, and registry pods to begin using the new certificate.

Author: Can we provide specific actions for completing step 3? Not familiar with how to do this myself.

Reviewer: kubectl rollout restart deployments -n < astronomer namespace >
kubectl rollout restart statefulsets -n < astronomer namespace >

          - Effect: Allow
            Action: route53:ListHostedZonesByName
            Resource: '*'
  AssumeRolePolicyDocument:
    Version: "2012-10-17"
    Statement:
      - Effect: "Allow"
        Principal:
          Service:
            - "ec2.amazonaws.com"
        Action:
          - "sts:AssumeRole"
```
For more information on how to complete this setup, refer to the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/create-node-role.html).

3. Create a "ClusterIssuer" resource that declares how requests for certificates will be fulfilled. To do so, first create the a `clusterissuer.yaml` file with the following values:
```yaml
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
    name: letsencrypt-prod
spec:
    acme:
        email: <your-email>
        server: https://acme-v02.api.letsencrypt.org/directory
        privateKeySecretRef:
            name: cert-manager-issuer-secret-key
        solvers:
        - selector: {}
          dns01:
            route53:
                region: <your-server-region>
```
Then, create the ClusterIssuer by running the following command:
```sh

> kubectl apply -f clusterissuer.yaml

Reviewer 3: ```suggestion
$ kubectl apply -f clusterissuer.yaml
```

```
4. Create a "Certificate" resource that declares the type of certificate you'll request from Let's Encrypt. To do so, first create the a `certificate.yaml` file, replacing `BASE_DOMAIN`:
```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
    name: acme-crt
spec:
    secretName: astronomer-tls
    dnsNames:
        - BASE_DOMAIN
        - app.BASE_DOMAIN
        - deployments.BASE_DOMAIN
        - registry.BASE_DOMAIN
        - houston.BASE_DOMAIN
        - grafana.BASE_DOMAIN
        - kibana.BASE_DOMAIN
        - install.BASE_DOMAIN
        - prometheus.BASE_DOMAIN
        - alertmanager.BASE_DOMAIN
    issuerRef:
        name: letsencrypt-prod
        kind: ClusterIssuer
        group: cert-manager.io
```
Then, create the cluster by running the following command and waiting a few minutes:
```sh

> kubectl apply -f certificate.yaml

Reviewer 3: ```suggestion
$ kubectl apply -f certificate.yaml
```

```
5. Ensure that the certificate was created by running:
```sh

> kubectl get certificates

Reviewer 3: ```suggestion
$ kubectl get certificates
```

```

## Manually Renew TLS certificates

Larger organizations with dedicated security teams will likely have their own processes for requesting and renewing TLS certificates. Regardless, there are specific steps you have to complete for Astronomer when renewing TLS certificates:

1. Delete your current TLS certificate by running the following command:
```sh

> kubectl delete secret astronomer-tls

Reviewer 3: ```suggestion
$ kubectl delete secret astronomer-tls
```

```
2. Follow the instructions for requesting an TLS certificate from your organization's security team as described in [Step 4: Configure TLS](https://www.astronomer.io/docs/enterprise/stable/install/aws/install-aws-standard#step-4-configure-tls). The linked guide is for setting up with AWS, but this step is the same regardless of which service you use.
3. Restart your Houston, nginx, and registry pods to begin using the new certificate by running the following commands:
```sh

> kubectl rollout restart deployments -n <your-namespace>

Reviewer 3: ```suggestion
$ kubectl rollout restart deployments -n <your-namespace>
```

> kubectl rollout restart statefulsets -n <your-namespace>

Reviewer 3: ```suggestion
$ kubectl rollout restart statefulsets -n <your-namespace>
```

---
title: "Renew TLS Certificates on Astronomer Enterprise"
navTitle: "Renew a TLS Certificate"
description: "How to update and auto-renew your organization's TLS Certificate for Astronomer"
---

## Overview

Once you set up an TLS certificate for your Astronomer platform, you'll need to establish a process for periodically renewing the certificate. This can be done in one of two ways:

* **Automatic renewal**: Let's Encrypt provides a service which automatically renews your TLS certificate every 90 days. We recommend this option for smaller organizations where your DNS administrator and cluster administrator are either the same person or on the same team.
* **Manual renewal**: Manual renewal looks very similar to the initial certificate creation process, except you replace your existing current certificate by creating a new certificate. We recommend this method for large organizations that have their own processes for issuing certificates.

## Automatically Renew TLS Certificates Using Let's Encrypt

> Let's Encrypt is a Certificate Authority that provides free, 90-day certificates using the ACME protocol. You can use the Cert Manager project for Kubernetes to automatically renew certificates.

Reviewer 3: ```suggestion
[Let's Encrypt](https://letsencrypt.org/) is a Certificate Authority that provides free, 90-day certificates using the ACME protocol. You can use the Cert Manager project for Kubernetes to automatically renew certificates.
```
